src/Ledger/Ledger/Properties.agda

```diff
@@ -82,29 +102,579 @@ instance
+          dp-c-∈-utxoDep = subst (λ x₂ → (x₂ , c) ∈ (utxoDeps ˢ))
+                                 (trans (sym a≡dp) a≡GAgaid) dpc∈utxoDeps
+
+          addr-dp-c-∈map : (returnAddr gast , GovActionDeposit gaid , c)
+                           ∈ mapˢ (returnAddr gast ,_) ( (utxoDeps ∣ ❴ GovActionDeposit gaid ❵)ˢ )
+          addr-dp-c-∈map =
+            to ∈-map $ (GovActionDeposit gaid , c)
+                     , (×-≡,≡→≡ (refl , refl) , res-singleton⁺{m = utxoDeps} dp-c-∈-utxoDep)
+        χ⊆χ' : χ utxoDeps ⊆ χ'
+        χ⊆χ' {a} x with (from ∈-map x)
+        ... | ((rwa , dp , c) , a≡dp , rwa-dp-c∈) with (from ∈-concatMapˢ rwa-dp-c∈)
+        ... | ((gaid , gast) , gaid-gast-∈-removed , rwa-dp-c-∈-map) with (from ∈-map rwa-dp-c-∈-map)
+        ... | ((dp' , _) , q≡ , q∈) = to ∈-map ((gaid , gast) , a≡GA , gaid-gast-∈-removed)
+          where
+          a≡GA : a ≡ GovActionDeposit gaid
+          a≡GA = let open ≡-Reasoning in begin
+            a                       ≡⟨ a≡dp ⟩
+            dp                      ≡⟨ cong (proj₁ ∘ proj₂) q≡ ⟩
+            dp'                     ≡⟨ cong proj₁ $ proj₂ (res-singleton''{m = utxoDeps} $ q∈) ⟩
+            GovActionDeposit gaid   ∎
+
+      main-invariance-lemma :
+          filterˢ isGADeposit (dom utxoDeps) ≡ᵉ fromList (map (λ (id , _) → GovActionDeposit id) govSt)
+          ----------------------------------------------------------------------------------------------------------
+        → filterˢ isGADeposit (dom utxoDeps') ≡ᵉ fromList (map (λ (id , _) → GovActionDeposit id) (filterᵇ Pᵇ govSt))
+
+      main-invariance-lemma (fst , snd) = ⇛ , ⇚
+        where
+        ⇛ : filterˢ isGADeposit (dom utxoDeps')
+            ⊆ fromList (map (λ (id , _) → GovActionDeposit id) (filterᵇ Pᵇ govSt))
+        ⇛ {a} h with from ∈-filter h
+        ... | (aGADep , a∈) with ∈-map⁻ (λ (id , _) → GovActionDeposit id)
+                                        $ from ∈-fromList (fst $ to ∈-filter (aGADep , res-comp-domᵐ a∈))
+        ... | ((tid , gast) , tidb∈govSt , refl) =
+          to ∈-fromList $ ∈ˡ-map-filter⁺{P? = λ u → ¿ P u ¿}
+                          ((tid , gast) , refl , tidb∈govSt , (res-comp-dom a∈) ∘ ∈-map⁺-∘)
+
+        ⇚ : fromList (map (λ (id , _) → GovActionDeposit id) (filterᵇ Pᵇ govSt))
+            ⊆ filterˢ isGADeposit (dom utxoDeps')
+        ⇚ {a} h with ∈-map⁻ (λ (id , _) → GovActionDeposit id) (from ∈-fromList h)
+        ... | (aid×st , aid×st∈ , refl) with ∈-filter⁻ (λ u → ¿ P u ¿) aid×st∈
+        ... | (aid×st∈govSt , Paid×st) =
+          to ∈-filter (refl , ∈-resᶜ-dom⁺ (a∉χ' , (to dom∈ $ proj₂ (from ∈-filter a∈))))
+           where
+           a∈ : a ∈ filterˢ isGADeposit (dom utxoDeps)
+           a∈ = snd $ to ∈-fromList $ to ∈ˡ-map (aid×st , refl , aid×st∈govSt)
+
+           a∉χ' : a ∉ χ'
+           a∉χ' a∈χ' with ∈-map⁻' a∈χ'
+           ... | q , refl , q∈rem = Paid×st (to ∈-map (q , refl , q∈rem))
```

Comment:
The same here, many of those rewriting steps are bidirectional.

Reply:
I did my best to make this a bidirectional proof and separated out the one step I couldn't make bidirectional into a separate lemma, called `map-filter-decomp` (which now appears just above the `main-invariance-lemma`).